rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+.. TODO::
+
+    Extend conceptual overview to describe how this will work with Services and Actions.
+    Services, since they are not sensitive to the many-to-many (many publisher) issue, unlike Topics, and because they do not have as many QoS settings that apply to them, they can probably have transfer functions that are plugins, rather than separate component nodes that repeat the service call, like the ros1_bridge.
+    Actions will be a combination of topics and services, but will have other considerations in the tooling.
+
+In order to support this vision, three missing features will need to be added into ROS 2 (which were also mentioned in the introduction):
+
+- enforcing type compatibility by version
+- providing access to the complete type description of types being used
+- making it possible to publish and subscribe to topics using just the type description
+
+These features are described in the following sections.
+
+Type Version Enforcement
+------------------------
+
+In order to detect type version mismatches and enforce them, a way to uniquely identify versions is required, and this proposal uses type version hashes.
+
+Type Version Hash
+^^^^^^^^^^^^^^^^^
+
+The type version hashes are not sequential and do not imply any rank among versions of the type.
+That is, given two version hashes of a type, there is no way to tell which is "newer".
+
+The type version hash can only be used to determine if type versions are equal and if there exists a chain of transfer functions that can convert between them.
+Because of this, when a change to a type is made, it may or may not be necessary to write transfer functions in both directions depending on how the interface is used.
+
+The type version hashes are calculated in a stable way and are not sensitive to trivial changes like changes in the comments or whitespace of the IDL file.
+The IDL file given by the user, which may be a ``.msg`` file, ``.idl`` file, or something else, is parsed and stored into a data structure which excludes things like comments but includes things that impact compatibility on the wire.
+
+The data structure includes:
+
+- a list of field names and types, but not default values
+- a recursive list of field names and types for referenced types
+- an optional user-defined interface version, or 0 if not provided
+
+.. TODO::
+
+    Should the message name, including package and namespace, be part of this?
+    Consider a situation where you have the same data structure but two different type names, should those two instances have the same hash?
+    They are unique when paired with their type name, so it should be ok, but it is a bit weird, perhaps, since we're not using this type hash to check for wire compatibility between differently named types.
+
+.. TODO::
+
+    Related TODO, should we just use the TypeDescription described in a later section?
```

Comment:
I think yes. Has there been further discussion on this point, as to whether the has should be based on anything other than the TypeDescription? It seems like TypeDescription peels away the unrelated info (comments, default values, etc.) leaving only the relevant information to hash.